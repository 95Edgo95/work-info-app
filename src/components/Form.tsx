import validator from "services/validator";
import Button from "components/Button";
import Input from "components/Input";
import * as React from "react";
import {Map} from "immutable";

import * as styles from "assets/css/form.css";

interface FormProps {
  attemptUpdateEntity: (data: any, id: string) => void;
  attemptCreateEntity: (data: any) => void;
  entity: Map<string, any>;
  entitiesSize: number;
  validations: any;
  inputs: any[];
  history: any;
}

let mounted = false;

export default function Form(props: FormProps) {
  const {
    entitiesSize,
    attemptCreateEntity,
    attemptUpdateEntity,
    history,
    entity,
    validations,
    inputs,
  } = props;

  let setStateMap = {};
  let fields = {};

  inputs.forEach(({ name, defaultValue }) => {
    const [field, setField]  = React.useState(entity ? entity.get(name) : defaultValue);

    setStateMap[name] = setField;
    fields[name] = field;
  });

  const [errors, setErrors] = React.useState<{ [key: string]: string }>({});

  function handleSubmit(event) {
    event.preventDefault();

    const nextErrors = validator(fields, validations);

    if (Object.keys(nextErrors).length > 0) {
      setErrors(nextErrors);
    } else {
      if (Object.keys(errors).length > 0) {
        setErrors({});
      }

      if (entity.get("id")) {
        attemptUpdateEntity(fields, entity.get("id"));
      } else {
        attemptCreateEntity(fields);
      }
    }
  }

  function handleChange({currentTarget: {name, value}}) {
    setStateMap[name](value);
  }

  React.useEffect(() => {
    if (mounted) {
      history.push("/");
    }
  }, [entitiesSize]);

  if (entity.get("id")) {
    entity.valueSeq().forEach((value) => {
      React.useEffect(() => {
        if (mounted) {
          history.push("/");
        }
      }, [value]);
    });
  }

  React.useEffect(() => {
    mounted = true;

    return () => {
      mounted = false;
    }
  }, []);

  return (
    <form onSubmit={handleSubmit} className={styles.form}>
      {inputs.map(({ placeholder, name, type }) => (
        <Input
          placeholder={placeholder}
          onChange={handleChange}
          error={errors[name]}
          value={fields[name]}
          name={name}
          type={type}
          key={name}
        />
      ))}
      <Button type="submit" className={styles.submitButton}>
        {(entity && entity.get("id")) ? "Update" : "Create"}
      </Button>
    </form>
  );
}
