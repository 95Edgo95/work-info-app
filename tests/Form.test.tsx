import Form from "../src/components/Form";
import { Map } from "immutable";
import { mount } from "enzyme";
import * as React from "react";

const clickFn = jest.fn();

describe('Form', () => {
  it('should render correctly', () => {
    const component = mount(
      <Form
        history={history}
        attemptCreateEntity={clickFn}
        attemptUpdateEntity={clickFn}
        entitiesSize={2}
        validations={{}}
        entity={Map()}
        inputs={[
          {
            placeholder: "First Name",
            name: "firstName",
            defaultValue: "",
          },
          {
            placeholder: "Last Name",
            defaultValue: "",
            name: "lastName",
          },
          {
            placeholder: "Passport",
            defaultValue: "",
            name: "passport",
          },
          {
            placeholder: "Email",
            defaultValue: "",
            name: "email",
          },
          {
            placeholder: "Birth Date",
            name: "birthDate",
            defaultValue: "",
            type: "date"
          },
        ]}
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
