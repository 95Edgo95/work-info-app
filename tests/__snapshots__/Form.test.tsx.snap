// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<Form
  attemptCreateEntity={[MockFunction]}
  attemptUpdateEntity={[MockFunction]}
  entitiesSize={2}
  entity={Immutable.Map {}}
  history={History {}}
  inputs={
    Array [
      Object {
        "defaultValue": "",
        "name": "firstName",
        "placeholder": "First Name",
      },
      Object {
        "defaultValue": "",
        "name": "lastName",
        "placeholder": "Last Name",
      },
      Object {
        "defaultValue": "",
        "name": "passport",
        "placeholder": "Passport",
      },
      Object {
        "defaultValue": "",
        "name": "email",
        "placeholder": "Email",
      },
      Object {
        "defaultValue": "",
        "name": "birthDate",
        "placeholder": "Birth Date",
        "type": "date",
      },
    ]
  }
  validations={Object {}}
>
  <form
    onSubmit={[Function]}
  >
    <Input
      key="firstName"
      name="firstName"
      onChange={[Function]}
      placeholder="First Name"
    >
      <div>
        <div
          className=""
        >
          <input
            name="firstName"
            onChange={[Function]}
            placeholder="First Name"
            type="text"
          />
        </div>
      </div>
    </Input>
    <Input
      key="lastName"
      name="lastName"
      onChange={[Function]}
      placeholder="Last Name"
    >
      <div>
        <div
          className=""
        >
          <input
            name="lastName"
            onChange={[Function]}
            placeholder="Last Name"
            type="text"
          />
        </div>
      </div>
    </Input>
    <Input
      key="passport"
      name="passport"
      onChange={[Function]}
      placeholder="Passport"
    >
      <div>
        <div
          className=""
        >
          <input
            name="passport"
            onChange={[Function]}
            placeholder="Passport"
            type="text"
          />
        </div>
      </div>
    </Input>
    <Input
      key="email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
    >
      <div>
        <div
          className=""
        >
          <input
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="text"
          />
        </div>
      </div>
    </Input>
    <Input
      key="birthDate"
      name="birthDate"
      onChange={[Function]}
      placeholder="Birth Date"
      type="date"
    >
      <DayPickerInput
        classNames={
          Object {
            "container": "DayPickerInput",
            "overlay": "DayPickerInput-Overlay",
            "overlayWrapper": "DayPickerInput-OverlayWrapper",
          }
        }
        clickUnselectsDay={false}
        component="input"
        dayPickerProps={Object {}}
        format="DD/MM/YYYY"
        formatDate={[Function]}
        hideOnDayClick={true}
        inputProps={
          Object {
            "className": undefined,
            "name": "birthDate",
          }
        }
        keepFocus={true}
        onDayChange={[Function]}
        overlayComponent={[Function]}
        parseDate={[Function]}
        placeholder="Birth Date"
        showOverlay={false}
        value=""
      >
        <div
          className="DayPickerInput"
        >
          <input
            name="birthDate"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            placeholder="Birth Date"
            value=""
          />
        </div>
      </DayPickerInput>
    </Input>
    <Button
      type="submit"
    >
      <button
        className=""
        type="submit"
      >
        Update
      </button>
    </Button>
  </form>
</Form>
`;
